docs/reqcontext.rst

```diff
@@ -175,6 +200,45 @@ Certain parts of the test system might also temporarily create a request
 context without calling the before-request handlers.  Make sure to write
 your teardown-request handlers in a way that they will never fail.
 
+.. _ensuring-teardown-callbacks-succeed:
+
+Ensuring teardown callbacks succeed
+-----------------------------------
+
+.. admonition:: Watch Out
+
+    Teardown callbacks must never raise. If they do, teardown callbacks that
+    haven't run yet won't run, and request/app contexts won't pop, which will
+    cause memory/file leaks at best and context persisting into later requests
+    at worst.
+
+    Note that extensions can register teardown callbacks, and those callbacks
```

Comment:
Thanks for the feedback! Also, I know what you mean. I have to admit, I was pretty miffed when I wrote this last year :)

Though, I still think those criticisms have a grain of truth. Maybe the right solution is a code change rather than more documentation? Putting the app context and request context pops into `finally` blocks seems reasonable.

Additionally, maybe we should introduce a `handle_teardown_exception` function, with some sensible default behavior, like what I put in the recipe (logging an exception to the current app logger). Maybe the function should optionally return an error page as well. `handle_user_exception` and `handle_exception` can't be used since teardowns might have run, meaning things that would normally be present might be gone.

I'd want to discuss those with maintainers before trying to write it, though.